hpat/datatypes/hpat_pandas_series_functions.py

```diff
@@ -173,6 +173,114 @@ def hpat_pandas_series_shape_impl(self):
     return hpat_pandas_series_shape_impl
 
 
+@overload_method(SeriesType, 'nsmallest')
+def hpat_pandas_series_nsmallest(self, n=5, keep='first'):
+    """
+    Pandas Series method :meth:`pandas.Series.nsmallest` implementation.
+
+    .. only:: developer
+       Test: python -m hpat.runtests -k hpat.tests.test_series.TestSeries.test_series_nsmallest*
+
+    Parameters
+    ----------
+    self: :obj:`pandas.Series`
+        input series
+    n: :obj:`int`, default 5
+        Return this many ascending sorted values.
+    keep: :obj:`str`, default 'first'
+        When there are duplicate values that cannot all fit in a Series of n elements:
+        first : return the first n occurrences in order of appearance.
+        last : return the last n occurrences in reverse order of appearance.
+        all : keep all occurrences. This can result in a Series of size larger than n.
+        *unsupported*
+
+    Returns
+    -------
+    :obj:`scalar`
+         returns :obj:`series`
+    """
+
+    _func_name = 'Method nsmallest().'
+
+    if not isinstance(self, SeriesType):
+        raise TypingError('{} The object must be a pandas.series. Given: {}'.format(_func_name, self))
+
+    if not isinstance(n, (types.Omitted, int, types.Integer)):
+        raise TypingError('{} The object must be an integer. Given n: {}'.format(_func_name, n))
+
+    if not isinstance(keep, (types.Omitted, str, types.UnicodeType)):
+        raise TypingError('{} The object must be an unicode. Given keep: {}'.format(_func_name, keep))
+
+    def hpat_pandas_series_nsmallest_impl(self, n=5, keep='first'):
+        if keep != 'first':
+            raise TypingError('Method nsmallest(). Unsupported parameters. Given keep != \'first\'')
+
+        local_data = hpat.hiframes.api.nlargest_parallel(self._data, n, False, hpat.hiframes.series_kernels.lt_f)
+        local_index = []
+        for a in numpy.unique(local_data):
+            for indices in numpy.where(self._data == a):
+                local_index.extend(indices)
+        local_index = local_index[:n]
```

Comment:
@densmirn This looks inefficient as we call numpy.where on the self._data up to n times. 
I think better would be to use dict/set of unique values to look for and iterate over self._data only once.
If the self._data[i] is in that dict/set we can append i to a list of indexes where it was found. Then we will need to merge all such lists into one local_index and cut off first n elements as before. What do you think?